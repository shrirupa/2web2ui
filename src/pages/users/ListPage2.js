import React from 'react';
import { connect } from 'react-redux';
import TimeAgo from 'react-timeago';
import { Tag } from '@sparkpost/matchbox';

import ListPage from 'src/components/listPage/ListPage';
import { Users } from 'src/components/images';
import PageLink from 'src/components/pageLink/PageLink';
import { SubaccountTag } from 'src/components';
import User from './components/User';

import { listUsers, deleteUser } from 'src/actions/users';
import { selectUsers } from 'src/selectors/users';
import { hasSubaccounts } from 'src/selectors/subaccounts';

import { SUBACCOUNT_REPORTING_ROLE } from 'src/constants';
import { hasUiOption } from 'src/helpers/conditions/account';

const COLUMNS = [
  { label: 'User', sortKey: 'name' },
  { label: 'Role', sortKey: 'access' },
  { label: 'Two Factor Auth', sortKey: 'tfa_enabled' },
  { label: 'Last Login', sortKey: 'last_login' },
  null
];

const SUB_COLUMNS = [
  { label: 'User', sortKey: 'name', width: '40%' },
  { label: 'Role', sortKey: 'access', width: '11%' },
  { label: 'Subaccount', sortKey: 'subaccount_id', width: '15%' },
  { label: 'Two Factor Auth', sortKey: 'tfa_enabled', width: '8%' },
  { label: 'Last Login', sortKey: 'last_login', width: '14%' },
  null
];

const DEFAULT_SORT_COLUMN = 'name';

const FILTER_BOX = {
  keyMap: { role: 'access' },
  exampleModifiers: ['name', 'email', 'role'],
  itemToStringKeys: ['username', 'name', 'email']
};

function formatRole(role) {
  if (role === SUBACCOUNT_REPORTING_ROLE) {
    return 'reporting';
  }
  return role;
}

const renderDeleteWarning = ({ name }) => (
  <p>
    <span>User "</span>
    <span>{name}</span>
    <span>
      " will no longer be able to log in or access this SparkPost account.
      All API keys associated with this user will be transferred to you.
    </span>
  </p>
);

export class ListPage2 extends React.Component {
  emptyState = () => {
    const { currentUser, users } = this.props;

    return {
      show: users.length === 1,
      title: 'Invite Your Team to SparkPost',
      image: Users,
      content: <p>Manage your team's accounts and roles.</p>,
      secondaryAction: {
        Component: PageLink,
        content: 'Edit your user account',
        to: `/account/users/edit/${currentUser.username}`
      }
    };
  }

  renderRow = (user) => {
    const { hasSubaccounts, isSubAccountReportingLive } = this.props;
    const fields = [
      <User name={user.name} email={user.email} username={user.username} />,
      formatRole(user.access),
      user.tfa_enabled ? (
        <Tag color={'blue'}>Enabled</Tag>
      ) : (
        <Tag>Disabled</Tag>
      ),
      user.last_login ? (
        <TimeAgo date={user.last_login} live={false} />
      ) : (
        'Never'
      )
    ];

    if (isSubAccountReportingLive && hasSubaccounts) {
      fields.splice(2, 0, user.subaccount_id ? <SubaccountTag id={user.subaccount_id} /> : null);
    }

    return {
      fields,
      actions: {
        editRoute: `/account/users/edit/${user.username}`,
        deletable: !user.isCurrentUser
      }
    };
  }

  deleteUser = (user) => this.props.deleteUser(user.username);

  render() {
    const { error, listUsers, loading, users, hasSubaccounts, isSubAccountReportingLive } = this.props;
    const columns = isSubAccountReportingLive && hasSubaccounts ? SUB_COLUMNS : COLUMNS;

    const empty = this.emptyState();

    return (
      <ListPage
        noun='user'
        primaryAction={{ content: 'Invite User', to: '/account/users/create' }}
        error={error}
        loadItems={listUsers}
        loading={loading}
        items={users}
        columns={columns}
        defaultSortColumn={DEFAULT_SORT_COLUMN}
        empty={empty}
        renderRow={this.renderRow}
        onDelete={this.deleteUser}
        deleteWarning={renderDeleteWarning}
        filterBox={FILTER_BOX}
      />
    );
  }
}

export default connect(
  (state) => ({
    currentUser: state.currentUser,
    error: state.users.error,
    loading: state.users.loading,
    users: selectUsers(state),
    hasSubaccounts: hasSubaccounts(state),
    isSubAccountReportingLive: hasUiOption('subaccount_reporting')(state)
  }),
  {
    listUsers,
    deleteUser
  }
)(ListPage2);
