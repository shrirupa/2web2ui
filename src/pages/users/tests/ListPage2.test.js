import React from 'react';
import { shallow } from 'enzyme';

import { ListPage2 } from '../ListPage2';

import { currentUser, users } from './fixtures';

describe('Page: ListPage2', () => {
  let baseProps;
  beforeEach(() => {
    baseProps = {
      currentUser,
      error: null,
      loading: false,
      users,
      listUsers: jest.fn(),
      deleteUser: jest.fn(),
      hasSubaccounts: false,
      isSubAccountReportingLive: true
    };
  });

  const subject = (props) => shallow(<ListPage2 {...baseProps} {...props} />);
  const subjectRenderRowFn = (props) => subject(props).find('ListPage').prop('renderRow');

  it('renders a list page', () => {
    expect(subject()).toMatchSnapshot();
  });

  it('formats users for the table collection', () => {
    const renderRow = subjectRenderRowFn();
    const row = renderRow({ ...users[0], isCurrentUser: false });
    expect(row).toMatchSnapshot();
  });

  it('displays TFA enabled status', () => {
    const renderRow = subjectRenderRowFn();
    expect(renderRow({ ...users[0], tfa_enabled: true })).toMatchSnapshot();
  });

  it('displays last login', () => {
    const renderRow = subjectRenderRowFn();
    expect(renderRow({ ...users[0], last_login: new Date('2018-04-01T16:15:00.000Z') })).toMatchSnapshot();
  });

  it('displays subaccount info when the account uses them', () => {
    expect(subject({ hasSubaccounts: true }).find('ListPage').prop('columns')).toContainEqual(expect.objectContaining({
      label: 'Subaccount'
    }));
  });

  it('should not render subaccount info when the feature is disabled', () => {
    expect(subject({ hasSubaccounts: true, isSubAccountReportingLive: false }).find('ListPage').prop('columns')).not.toContainEqual(expect.objectContaining({
      label: 'Subaccount'
    }));
  });

  it('does not allow deletion of the current user', () => {
    const renderRow = subjectRenderRowFn();
    expect(renderRow({ ...users[0], isCurrentUser: true }).actions.deletable).toBeFalsy();
  });

  it('deletes users', () => {
    const deleteFn = subject().find('ListPage').prop('onDelete');
    deleteFn(users[0]);
    expect(baseProps.deleteUser).toHaveBeenCalledWith(users[0].username);
  });
});
