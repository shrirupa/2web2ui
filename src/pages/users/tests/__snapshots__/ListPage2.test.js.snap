// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page: ListPage2 displays TFA enabled status 1`] = `
Object {
  "actions": Object {
    "deletable": true,
    "editRoute": "/account/users/edit/test-user-1",
  },
  "fields": Array [
    <User
      email="user1@test.com"
      name="Test User 1"
      username="test-user-1"
    />,
    "admin",
    <Tag
      color="blue"
    >
      Enabled
    </Tag>,
    "Never",
  ],
}
`;

exports[`Page: ListPage2 displays last login 1`] = `
Object {
  "actions": Object {
    "deletable": true,
    "editRoute": "/account/users/edit/test-user-1",
  },
  "fields": Array [
    <User
      email="user1@test.com"
      name="Test User 1"
      username="test-user-1"
    />,
    "admin",
    <Tag>
      Disabled
    </Tag>,
    <TimeAgo
      component="time"
      date={2018-04-01T16:15:00.000Z}
      formatter={[Function]}
      live={false}
      maxPeriod={Infinity}
      minPeriod={0}
      now={[Function]}
    />,
  ],
}
`;

exports[`Page: ListPage2 formats users for the table collection 1`] = `
Object {
  "actions": Object {
    "deletable": true,
    "editRoute": "/account/users/edit/test-user-1",
  },
  "fields": Array [
    <User
      email="user1@test.com"
      name="Test User 1"
      username="test-user-1"
    />,
    "admin",
    <Tag>
      Disabled
    </Tag>,
    "Never",
  ],
}
`;

exports[`Page: ListPage2 renders a list page 1`] = `
<ListPage
  columns={
    Array [
      Object {
        "label": "User",
        "sortKey": "name",
      },
      Object {
        "label": "Role",
        "sortKey": "access",
      },
      Object {
        "label": "Two Factor Auth",
        "sortKey": "tfa_enabled",
      },
      Object {
        "label": "Last Login",
        "sortKey": "last_login",
      },
      null,
    ]
  }
  defaultSortColumn="name"
  deleteWarning={[Function]}
  empty={
    Object {
      "content": <p>
        Manage your team's accounts and roles.
      </p>,
      "image": [Function],
      "secondaryAction": Object {
        "Component": [Function],
        "content": "Edit your user account",
        "to": "/account/users/edit/test-user-1",
      },
      "show": false,
      "title": "Invite Your Team to SparkPost",
    }
  }
  error={null}
  filterBox={
    Object {
      "exampleModifiers": Array [
        "name",
        "email",
        "role",
      ],
      "itemToStringKeys": Array [
        "username",
        "name",
        "email",
      ],
      "keyMap": Object {
        "role": "access",
      },
    }
  }
  items={
    Array [
      Object {
        "access": "admin",
        "email": "user1@test.com",
        "name": "Test User 1",
        "tfa_enabled": false,
        "username": "test-user-1",
      },
      Object {
        "access": "admin",
        "email": "user2@test.com",
        "name": "Test User 2",
        "tfa_enabled": true,
        "username": "test-user-2",
      },
      Object {
        "access": "admin",
        "email": "user3@test.com",
        "name": "Test User 3",
        "subaccount_id": 125,
        "tfa_enabled": true,
        "username": "test-user-3",
      },
    ]
  }
  loadItems={[MockFunction]}
  loading={false}
  noun="user"
  onDelete={[Function]}
  primaryAction={
    Object {
      "content": "Invite User",
      "to": "/account/users/create",
    }
  }
  renderRow={[Function]}
/>
`;
